File: es14/es14.go
package main

import (
	"fmt"
	"os"
	"strconv"
)

func main() {
	rv := mainWithInt()
	if rv != 0 {
		os.Exit(rv)
	}
}

func mainWithInt() int {
	if len(os.Args) < 2 {
		fmt.Fprintf(os.Stderr, "Usage: %s <number>\n", os.Args[0])
		return -1
	}

	number, err := strconv.ParseInt(os.Args[1], 10, 32)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Cannot parse the number %s\n", os.Args[1])
		return -1
	}

	match := []int{}
	for _, c := range os.Args[1] {
		match = append(match, int(c)-'0')
	}
	mc := 0

	var should_terminate = func(digit int) bool {
		if digit == match[mc] {
			mc += 1
		} else if digit == match[0] {
			mc = 1
		} else {
			mc = 0
		}
		return mc == len(match)
	}

	recipes := []int{3, 7}
	e1 := 0
	e2 := 1

	for {
		r := recipes[e1] + recipes[e2]
		if r >= 10 {
			d := r / 10
			recipes = append(recipes, d)
			if should_terminate(d) {
				break
			}

			r -= 10
		}
		recipes = append(recipes, r)
		if should_terminate(r) {
			break
		}

		e1 = (e1 + 1 + recipes[e1]) % len(recipes)
		e2 = (e2 + 1 + recipes[e2]) % len(recipes)
	}

	if len(recipes) >= int(number)+10 {
		fmt.Printf("Solution1: ")
		for i := 0; i < 10; i++ {
			fmt.Printf("%d", recipes[int(number)+i])
		}
		fmt.Println()
	}

	fmt.Println("Solution2:", len(recipes)-len(match))

	return 0
}

File: es14/es14.py
#!/usr/bin/env python3

import sys

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} number".format(sys.argv[0]), file = sys.stderr)
        exit(-1)

    try:
        number = int(sys.argv[1])
    except:
        print("Cannot parse the number {}".format(sys.argv[1]), file = sys.stderr)
        exit(-1)

    recipes = [3, 7]
    e1 = 0
    e2 = 1

    match = [int(x) for x in sys.argv[1]]
    mc = 0

    def should_end(digit):
        global mc
        if digit == match[mc]:
            mc += 1
        elif digit == match[0]:
            mc = 1
        else:
            mc = 0

        return mc == len(match)

    it = 0
    while True:
        it += 1
        recipe = recipes[e1] + recipes[e2]
        if recipe >= 10:
            digit = recipe // 10
            recipes.append(digit)
            if should_end(digit):
                break

        digit = recipe % 10
        recipes.append(digit)
        if should_end(digit):
            break

        e1 = (e1 + recipes[e1] + 1) % len(recipes)
        e2 = (e2 + recipes[e2] + 1) % len(recipes)

    if len(recipes) >= number + 10:
        out = []
        for i in range(10):
            out.append(str(recipes[number+i]))

        print("Sequence: {}".format("".join(out)))

    if mc == len(match):
        print("Recipes on left: {}".format(len(recipes) - mc))

    print("Number of iterations: {}".format(it))
